### PR TITLE
Major refactoring for 0.2.0

### PR DESCRIPTION
**Pull request by vudaltsov**
# My most important considerations
1. Atomic types are now implemented as enums with a single case `self`. This should drastically decrease memory consumption and probably increase speed in the reflection package. As a bonus it allows to compare some types via `===`.
2. Removed `TypeAlias` and cancelled restrictions for implementing `Type`. I just thought that if one wants to implement `Type`, it will become obvious that `Visitor` will not support the custom type unless it uses one of the existing `Visitor::visit*` methods inside `accept`.
3. Rethought the concept of this package. I realized that it's just a low-level API for other packages, it shouldn't provide nice facades or care that much about naming suffixes. That is why I changed them to `Type` and removed developer-friendly logic from constructors of `Shape` and `Callable`. We can discuss a nice facade package later, it's not a high-priority task.
4. I have decided that this package should not provide any types as aliases. It's a responsibility of other tools to resolve that. So `NulalbleType`, `PositiveIntType`, `ScalarType`, etc. are now provided as independent types. I think we'll soon publish a separate package that will be able to optimize/flatten types and resolve `?(scalar)|0|positive-int` to `null|true|false|int|float|string` for further faster manipulations.
5. According to 4. added `NegativeInt`, `NonNegativeInt`, `NonPositiveInt` ([supported by Psalm 5](https://psalm.dev/docs/annotating_code/type_syntax/scalar_types/)).

# Questions
- Does it all make sense? :)
- Can we find a better name for case `self`?

### CODE REVIEW

File: src/Type/GeneratorT.php [hunk 1]
```diff
@@ -1,46 +0,0 @@
-<?php
```
Comment (Etherlord) [inline src/Type/GeneratorT.php:1]: Why was the Generator type removed?

**Comment by aivchen** [comment]
>Atomic types are now implemented as enums with a single case self.

1. In my opinion It looks like a trick. Enums are for defining a set of named constants. And when we define only one constant in enum and we know that we will never add another constant into it, we do not use enum for its intended purpose.
2. This breaks the uniformity of types instantiation. And every time you need to instantiate some type you have to think is it enum or class:
  ```php
  IntType::self
  new IntRangeType(...)
  ``` 
3. I'm not sure is it possible, but may be, in future, some types may introduce parameters. In this case you have to change enum to class anyway.

Everything else looks great I think.

**Comment by vudaltsov** [comment]
@aivchen ,
1. it is actually a trick! But an awesome trick that fixes this issue: https://github.com/vimeo/psalm/discussions/9328
2. Yeah, I thought about that, it is very inconvenient. I'll try to come up with another trick :)
3. I think that even if it's possible, it might be easily changed in a major release.

**Comment by vudaltsov** [comment]
See https://github.com/php-extended-type-system/type/commit/2e2446a4cdcc81f75c8184f9a9f232c5a78cd973.

**Pull request closed without merging**